iwyu.cc

```diff
@@ -2423,11 +2423,18 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     // in the (common) case the placement arg looks like '&something'.
     // (This is possibly wrong for classes that override operator&, but
     // those classes deserve what they get.)
-    if (!expr->getOperatorNew() &&
```

Comment:
Doesn't the comment above need updating?

Reply:
Hehe, I thought I did...? Are you missing something in particular? I'll look it over and try to edit it down a bit anyway, it's pretty wordy.

Reply:
Rewrote the comment, let me know what you think.

Reply:
It looks better, thanks! Btw, I'm not an expert in `operator new` overloading and cannot say whether the new logic is optimal tradeoff or not.

Reply:
I think it's good enough -- there is a loophole where:
```
template<class T>
T* construct(T& target) {
  return new(std::addressof(target)) T;
}
```
works as expected, but stuff like:
```
template<class T>
std_align_val_t& getAlign<T>() {
  return std_align_val_t(T::getPreferredAlignment());
}

template<class T>
T* construct() {
  return new(getAlign<T>()) T;
}
```
will falsely be marked as placement-new, needing `#include <new>` (on the other hand `std::align_val_t` is in `<new>`, so the result is still valid, just that diagnose the placement-new as also needing `<new>`).